[bitnami/spark] Open up https if security.ssl.enabled

brsolomon-deloitte opened the pull request:
- Adds https containerPort on master, default .Values.master.webPortHttps -> 8480
- Adds https containerPort on workers, default .Values.worker.webPortHttps -> 8481
- Adds port to svc-master.yaml, default 443, pointing to spark master https
- Fixes #8417
- Fixes #8446

migruiz4 commented on bitnami/spark/templates/statefulset-master.yaml:
```diff
@@ -145,6 +145,10 @@ spec:
             - name: http
               containerPort: {{ .Values.master.webPort }}
               protocol: TCP
+            {{- if .Values.security.ssl.enabled }}
+            - name: https
+              containerPort: {{ .Values.master.webPortHttps }}
+            {{- end }}
```
For consistency, please add the `protocol` field as you did in the worker statefulset and as in the field above.
```suggestion
            {{- if .Values.security.ssl.enabled }}
            - name: https
              containerPort: {{ .Values.master.webPortHttps }}
              protocol: TCP
            {{- end }}
```

brsolomon-deloitte commented:
Verified that `curl -k https://spark-master-svc` will work within the cluster with this change and responds with web UI html.

```
$ curl -I -k https://spark-master-svc
HTTP/1.1 200 OK
Date: Fri, 17 Dec 2021 17:02:21 GMT
Cache-Control: no-cache, no-store, must-revalidate
X-Frame-Options: SAMEORIGIN
X-XSS-Protection: 1; mode=block
X-Content-Type-Options: nosniff
Content-Type: text/html;charset=utf-8
Content-Length: 7300
Server: Jetty(9.4.43.v20210629)
```

migruiz4 commented:
Hi @brsolomon-deloitte,

Thank you very much for your contribution! I have just replied at https://github.com/bitnami/charts/issues/8417 when I saw this PR.

Just a minor comment in my change request plus an additional requirement.

This change requires modifying the Spark image, since modifying the value `webPortHttps` will not reflect in the Spark configuration.

Adding an additional env variable similar to `SPARK_WEBUI_PORT`, for example, `SPARK_WEBUI_SSL_PORT` and modifying this section of the container logic would be required: https://github.com/bitnami/bitnami-docker-spark/blob/cc41d1caa11c607c244653bafd868239c0f34f36/3/debian-10/rootfs/opt/bitnami/scripts/libspark.sh#L229-L245

brsolomon-deloitte commented:
@migruiz4 

> This change requires modifying the Spark image, since modifying the value webPortHttps will not reflect in the Spark configuration.

To be clear, you are suggesting to set `spark.ssl.port`  within the Bash function `spark_enable_ssl`, based on value of `SPARK_WEBUI_SSL_PORT` ?

brsolomon-deloitte commented:
Maybe I am doing something wrong here, but it seems that the `ca.crt` isn't signed for `spark-master-svc` if `security.ssl.autoGenerated=true`.

```
$ k get secret -n xxxxx spark-crt -o jsonpath='{.data.ca\.crt}' | base64 -d > spark-ca.crt
$ openssl x509 -noout -text -in spark-ca.crt 
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            1c:86:e8:1d:f7:8d:77:0b:ec:72:02:1e:e7:a9:9e:a5
    Signature Algorithm: sha256WithRSAEncryption
        Issuer: CN=spark-internal-ca
        Validity
            Not Before: Dec 17 16:53:04 2021 GMT
            Not After : Dec 17 16:53:04 2022 GMT
        Subject: CN=spark-internal-ca
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                Public-Key: (2048 bit)
                Modulus:
                    00:c5:cf:26:e0:1d:08:a2:bf:17:f2:28:94:86:ed:
                    6c:5f:68:6f:68:57:be:c4:73:dd:1a:00:e5:df:cd:
                    71:7e:08:93:cf:e8:ac:0f:3e:78:b3:ae:4d:36:99:
                    26:ea:22:c3:c0:a1:62:26:81:42:73:6a:aa:4a:66:
                    fe:56:53:40:41:09:6e:b3:58:ea:8c:f3:15:da:0f:
                    6d:3e:36:1d:18:4d:0e:eb:62:72:49:fd:c5:1e:e1:
                    91:df:28:48:ba:8a:09:5e:af:84:11:17:3e:49:f9:
                    59:f7:e5:90:c3:e8:05:4f:2e:39:d7:c5:a4:a8:73:
                    60:43:b3:48:bb:6c:76:89:45:b7:09:78:d4:d9:57:
                    59:59:37:bf:1b:ae:b6:ff:04:ad:ad:f0:0d:d5:a8:
                    7a:17:bb:e9:52:19:00:f5:2f:b2:da:9f:b6:1e:ab:
                    d5:87:8b:0a:40:d6:36:24:da:d1:41:67:3d:fb:92:
                    0d:44:1b:cb:3a:73:20:d7:05:04:c4:e5:68:a9:ab:
                    aa:0f:5d:58:08:87:df:ed:74:c3:1f:e7:c5:42:df:
                    f5:6b:bf:88:76:07:64:56:4b:c0:94:92:61:47:4f:
                    fd:7f:99:c1:9c:61:b2:9e:75:2c:63:19:7e:50:e1:
                    20:d5:fd:91:98:ef:c5:b2:19:9b:0c:8f:fa:83:b8:
                    42:a1
                Exponent: 65537 (0x10001)
        X509v3 extensions:
            X509v3 Key Usage: critical
                Digital Signature, Key Encipherment, Certificate Sign
            X509v3 Extended Key Usage: 
                TLS Web Server Authentication, TLS Web Client Authentication
            X509v3 Basic Constraints: critical
                CA:TRUE
            X509v3 Subject Key Identifier: 
                E4:09:92:D1:9E:31:64:46:99:7B:4C:4F:BB:C9:65:0E:DA:48:C7:BE
    Signature Algorithm: sha256WithRSAEncryption
         28:dd:07:fc:49:db:8a:f3:13:5c:9c:bc:cb:68:5f:2f:7d:66:
         7f:83:ec:21:5c:3a:f0:e9:48:e4:4a:d8:fd:05:1d:64:be:1d:
         4c:f1:4a:1c:d5:ca:4e:10:59:72:8f:d2:d9:ff:01:d4:0e:b3:
         37:73:fc:0a:9c:6f:e7:a8:8b:01:c0:92:47:66:e6:93:21:ca:
         38:3b:b0:68:b6:21:a6:80:72:fe:cf:f7:fa:8e:a8:72:68:20:
         4e:f7:f1:04:e2:ee:a8:d4:79:2c:74:9e:a0:a4:0a:15:8b:68:
         66:9e:54:3a:f6:4d:05:9b:90:3a:e0:a3:ff:d1:54:e0:ef:3d:
         0d:53:aa:19:b4:c3:47:0a:eb:01:a9:11:0e:e3:f5:08:bf:18:
         1c:4e:89:42:bb:3d:c0:41:d7:3d:64:11:b1:9a:76:ac:41:8a:
         4d:23:2b:4b:cd:57:bc:54:d8:58:fb:fe:53:01:43:99:cc:d2:
         f2:17:d2:04:03:5d:08:f0:69:c7:d5:88:60:47:02:e7:2c:f3:
         9e:ff:ae:53:23:1f:b2:57:6a:18:6b:84:b2:a8:79:07:1e:7b:
         6d:5a:3c:be:0e:0f:77:3d:e1:d4:2c:0f:1e:56:48:2a:c4:05:
         ba:c8:78:1a:be:da:a8:94:e6:1a:ac:1f:ff:53:3d:86:ba:9b:
         23:8b:66:6a
```

javsalgar commented:
Just a note to let you know that @migruiz4 is on PTO. As soon as he comes back he will provide you with feedback.

migruiz4 commented:
Hi @brsolomon-deloitte,

Sorry for the late response. Answering your questions:

> To be clear, you are suggesting to set spark.ssl.port within the Bash function spark_enable_ssl, based on the value of SPARK_WEBUI_SSL_PORT ?

Yes, setting `spark.ssl.port` in the bash function based on the value of a new env variable (e.g. `SPARK_WEBUI_SSL_PORT`) should allow setting the env variable.
 
> EDIT: based on charts/bitnami/spark/templates/tls-secrets.yaml, it looks like this is signed for the headless service only. Would it make sense to add the spark-master-svc to alternate names?

I agree it could make sense to add `spark-master-svc` to the list of alternative names, but I would like to point something out regarding it. 
Using certificates created by the `autogenerated` option may not be recommended for production since it is intended to make TLS/SSL settings easier, and better practice would require creating certificates with the IPs/Alternative names you intend to expose.

brsolomon-deloitte commented:
Added a PR at https://github.com/bitnami/bitnami-docker-spark/pull/64.

> Using certificates created by the autogenerated option may not be recommended for production since it is intended to make TLS/SSL settings easier, and better practice would require creating certificates with the IPs/Alternative names you intend to expose.

I am in agreement with this, but also of the opinion that, if `autogenerated` is offered as an option, then it should work fully. Given this, I think it is reasonable to add spark-master-svc to the list of alternate names.

brsolomon-deloitte commented:
@migruiz4 Hi, I think this one could benefit from another look when you have a chance. I have:

- Opened a PR to the Docker repo
- Added spark-master-svc to alternative names in the cert
- Cleaned up lint issues

migruiz4 commented on bitnami/spark/templates/statefulset-worker.yaml:
```diff
@@ -249,6 +254,8 @@ spec:
             - name: SPARK_SSL_PROTOCOL
               value: {{ .Values.security.ssl.protocol | quote }}
+            - name: SPARK_WEBUI_SSL_PORT
+              value: {{ .Values.worker.webPortHttps }}
```
Missing quotes will cause the Statefulset to fail
```suggestion
              value: {{ .Values.worker.webPortHttps | quote }}
```

migruiz4 commented on bitnami/spark/templates/statefulset-master.yaml:
```diff
@@ -226,6 +231,8 @@ spec:
             - name: SPARK_SSL_PROTOCOL
               value: {{ .Values.security.ssl.protocol | quote }}
+            - name: SPARK_WEBUI_SSL_PORT
+              value: {{ .Values.master.webPortHttps }}
```
Missing quotes will cause the Statefulset to fail
```suggestion
              value: {{ .Values.master.webPortHttps | quote }}
```

migruiz4 commented on bitnami/spark/templates/statefulset-master.yaml:
```diff
@@ -149,6 +149,11 @@ spec:
             {{- end }}
               containerPort: {{ .Values.master.webPort }}
               protocol: TCP
+            {{- if .Values.security.ssl.enabled }}
+            - name: https
+              containerPort: {{ .Values.master.webPortHttps }}
+              protocol: TCP
+            {{- end }}
```
Same issue as the Service. This causes `https` port to be duplicated.

In this case I would suggest:
```
            {{- if not .Values.security.ssl.enabled }}
            - name: http
              containerPort: {{ .Values.master.webPort }}
              protocol: TCP
            {{- else }}
            - name: https
              containerPort: {{ .Values.master.webPortHttps }}
              protocol: TCP
            {{- end }}
```

brsolomon-deloitte replied:
Changed in d228c74ef

migruiz4 commented on bitnami/spark/templates/statefulset-worker.yaml:
```diff
@@ -147,6 +147,11 @@ spec:
             - name: http
               containerPort: {{ .Values.worker.webPort }}
               protocol: TCP
+            {{- if .Values.security.ssl.enabled }}
+            - name: https
+              containerPort: {{ .Values.worker.webPortHttps }}
+              protocol: TCP
+            {{- end }}
```
Please use the same logic as the master Statefulset.

```suggestion
            {{- if not .Values.security.ssl.enabled }}
            - name: http
              containerPort: {{ .Values.worker.webPort }}
              protocol: TCP
            {{- else }}
            - name: https
              containerPort: {{ .Values.worker.webPortHttps }}
              protocol: TCP
            {{- end }}
```

migruiz4 reviewed (changes requested):
Hi @brsolomon-deloitte,

Thank you for all the changes applied. Please have a look at my comments below with the issues I found while testing this PR.

brsolomon-deloitte commented on bitnami/spark/templates/svc-master.yaml:
```diff
@@ -41,5 +41,16 @@ spec:
+      {{- else if eq .Values.service.type "ClusterIP" }}
+      nodePort: null
+      {{- end }}
+    {{- end }}
```
Changed in e6066bdab. Sorry, that was a dumb mistake on my part.

brsolomon-deloitte commented:
@migruiz4 one thing I want to confirm here: are we sure we want to use this "either/or" logic to serve either http or https, but not both?

The logs show that Spark serves over both, rather than just one or the other exclusively.

```
21/12/17 16:00:41 INFO Utils: Successfully started service 'MasterUI (HTTPS)' on port 8480.
21/12/17 16:00:41 INFO Utils: Successfully started service 'MasterUI' on port 8080.

21/12/17 16:03:07 INFO Utils: Successfully started service 'WorkerUI (HTTPS)' on port 8481.
21/12/17 16:03:07 INFO Utils: Successfully started service 'WorkerUI' on port 8081.
```

migruiz4 commented:
Hi @brsolomon-deloitte,

Thank you very much for applying the suggested changes! Your PRs look good to me, but I need to ask you to sign your commits before we can merge this PR.

For more information, please check the following link https://github.com/bitnami/charts/pull/8447/checks?check_run_id=4910317855

Regarding your question, although Spark exposes both HTTP and HTTPS, I'm fine with the chart only exposing HTTPS when it is configured.

If I were a user deploying Spark with security features, I would expect it to expose just HTTPS endpoints and not both. 
Then, in case of users asked for this feature, I would consider adding an extra value hideHttp yes/no to control that behavior.

brsolomon-deloitte commented:
Due to messiness with rebasing + signing old commits I have opened https://github.com/bitnami/charts/pull/8844 which supercedes this but has the same diff.

migruiz4 commented:
Thank you very much for your contribution @brsolomon-deloitte! I will now close this PR since the changes were merged at #8844
